builder/common/ssm/session.go

```diff
@@ -84,21 +86,48 @@ func (s Session) getCommand(ctx context.Context) ([]string, string, error) {
 	return args, *session.SessionId, nil
 }
 
+// terminate an interactive Systems Manager session with a remote instance via the
+// AWS session-manager-plugin. Session cannot be resumed after termination.
+func (s Session) terminateSession(sessionID string, ui packersdk.Ui) {
+	log.Printf("ssm: Terminating PortForwarding session %q", sessionID)
+	_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
+	if err != nil {
+		ui.Error(fmt.Sprintf("Error terminating SSM Session %q, this does not affect the built AMI. Please terminate the session manually: %s", sessionID, err))
+	}
+}
+
 // Start an interactive Systems Manager session with a remote instance via the
 // AWS session-manager-plugin. To terminate the session you must cancell the
 // context. If you do not wish to terminate the session manually: calling
 // StopSession on a instance of this driver will terminate the active session
 // created from calling StartSession.
 // To stop the session you must cancel the context.
 func (s Session) Start(ctx context.Context, ui packersdk.Ui, sessionChan chan struct{}) error {
-	for ctx.Err() == nil {
+	exitSession := false
+	for ctx.Err() == nil && !exitSession {
 		log.Printf("ssm: Starting PortForwarding session to instance %s", s.InstanceID)
 		args, sessionID, err := s.getCommand(ctx)
 		if sessionID != "" {
 			defer func() {
-				_, err := s.SvcClient.TerminateSession(&ssm.TerminateSessionInput{SessionId: aws.String(sessionID)})
-				if err != nil {
-					ui.Error(fmt.Sprintf("Error terminating SSM Session %q. Please terminate the session manually: %s", sessionID, err))
+				s.terminateSession(sessionID, ui)
+			}()
+			// If the instance is terminated the session must be terminated as well
+			// Otherwise the start-session command will exit with status -1
+			go func() {
```

Comment:
Seeing this is in a loop, this coroutine will be spawned multiple times, but from my understanding, it should only be invoked once, right?
The deferred function above as well I'd think, since we only kill the session when we leave the parent function.

Reply:
You're right. Does the following approach make sense?

I can create a sessionIds slice, this is used to keep record of sessions created. because it's a loop we might create more than one session. 

when the instance is in termination the session must be immediately terminated, and we cannot rely on the defer([same as this comment](https://github.com/hashicorp/packer-plugin-amazon/pull/324/files#r1103209580)). This is where we need those sessionIds to terminate all of them.

This second approach only has one coroutine and that one will terminate all the sessions created at once.

Reply:
One more thing, if we do the termination immediately then the defer is redundant.

Reply:
Regarding this thread, I misunderstood the intent here, my bad, I thought this was only one session ID, but from the looks of it, there are indeed multiple sessions created, one per loop.

In this case, I don't think we'll benefit from storing the session IDs and purging them after we're done, so I think we can continue with the proposed approach.

Reply:
I also agree with your sentiment, since we terminate the session here, we can remove the deferred function entierly.